### DISCUSSION TITLE
predict

### DISCUSSION THREAD

**Discussion by AlainZuur** [JenniNiku/gllvm#160] [General]
Second question of the day (sorry..I stack them)

This one has been nagging me for a while. Suppose that I want to visualise the model fit. Say for a Poisson GLLVM.

In an ordinary Poisson GLM, you fit the model, create some made-up values for your covariates, use predict with type = "link" to get the predicted values and a 95% CI, and then apply the exponential function. Like this:
1. Fit model: M1 <- glm(Y ~ X + Z, family = "poisson")
2. MyData <- data.frame(X = c(1,2,3),  Z = c(4,5,6))
3. P <- predict(M1, newdata = MyData, type = "link")
4. Get 95% CIs
5. Apply the exponential function.
6. Plot the MyData data versus the predicted values...and you have your model fit with a 95% CI.

Now a Poisson GLMM. Say..with a random effect Site. When you want to do the same thing as above, then most people just predict by setting the random effects to 0. The conditional effect. And most people then call this "the effects of the covariates for a typical, or average site. I do that too. But what is an 'average' site. 

If I apply the predict function to a GLLVM with new data, then I also get predicted values. They show me the effect of the covariates....while setting the latent variables to 0...and the random effects to 0. So..I have the same problem as above.

Nice explanation here for conditional versus marginal effects:
https://www.andrewheiss.com/blog/2022/11/29/conditional-marginal-marginaleffects/

My question is essentially: All that simulation stuff that Andrew Heiss is doing to get marginal effects..those 4 different approaches......we can do that for GLLVM as well? 

A GLLVM is more complicated...there are row-specific random effects and latent variables (and perhaps more depending on the model)....but everything we need is in:

#'      - mean values:       MyModel$TMBfn$par[MyModel$Hess$incl] 
#'      - covariance matrix: MyModel$Hess$cov.mat.mod

So....everything is simulatable.

And such a simulation would give the marginal effects plus a measure of uncertainty around those predicted values.

Maybe I should just try? 

Alain

**Comment by BertvanderVeen** [comment]
I am not familiar with Andrew Heiss, or his blog, and do not tend to think a lot about these kind of post-hoc results (though I probably should do it more). GLLVMs are just big fat GLMMs in most cases, so if you can do it with a GLMM you can do it with a GLLVM. It just might be a tad more tedious.

> A GLLVM is more complicated...there are row-specific random effects and latent variables (and perhaps more depending on the model)....but everything we need is in:

> #' - mean values: MyModel$TMBfn$par[MyModel$Hess$incl]
> #' - covariance matrix: MyModel$Hess$cov.mat.mod

Yup. That is, everything for fixed effects, not for random effects. The joint precision matrix is also stored by default only for VA, under `MyModel$TMBfn$Hess.full`. Though it would be possible to get the joint precision matrix for models fitted by LA via TMB's sdreport.

On this see also the discussion in #24 and @jebyrnes his https://gist.github.com/jebyrnes/a60ec172a391ed8dd3d44cbe1eed0bed.
